### ISSUE TITLE
Add "labels" map for to target filter

### ISSUE THREAD

**Issue by martin-schaefer** [promregator/promregator#110]
As discussed in #101 I suggest to add support for selecting apps by label. A use case example is shown below, where promregator would select all apps in a given space which have the label "metrics-provider-type" with value "spring-boot". The attribute "labels" would be a map.

```yaml
     targets:
       - orgName: my-org
         spaceName: my-space
         labels: 
           metrics-provider-type: spring-boot
         path: actuator/prometheus
```
This would allow to tag the apps whould should be selected by promregator, What do you think?

**Comment by eloo** [comment]
@martin-schaefer we had already a similar discussion here #61 
in general i also support your idea, but i don't know how complex it is to implement.
also i'm curious how you want to set the label on cf side on the app? AFAIK there isn't an option to set a label to an app?

**Comment by martin-schaefer** [comment]
It is possible:   https://docs.cloudfoundry.org/adminguide/metadata.html
 Youcan even query the CF API by labels so it should not be a big deal.

**Comment by eloo** [comment]
Ah okay. yeah thats interesting

**Comment by eaglerainbow** [comment]
Copying here two further arguments pro labels, which were raised by @martin-schaefer in #101:

> Using app names has the disadvantage to required a name-pattern or to add every's apps name to promregator configuration. This assumes a scenario, where only selected apps should be picked by promregator.

> A regex for labels would not be necessary because one is free to define specialized labels like "prometheus-metrices-provider" and can use exact match.

**Comment by eaglerainbow** [comment]
> in general i also support your idea, but i don't know how complex it is to implement.

In general, I am on a very similar state on this matter: Also our previous "service broker"/"User-provided service" approach looked very promising, but the details ended up with so many still-open questions of which most are beyond hope of being addressed (see #61). Yet, I would be willing to give the label's idea a try. However, we already have one con with it: Apparently, there is no CF CLI support for it, but you must `cf curl` it manually. Or, @martin-schaefer, are you aware of any plugin or something like this?

**Comment by eaglerainbow** [comment]
> This would allow to tag the apps which should be selected by promregator, What do you think?

I have a little trouble with your configuration proposal. Why do you want to have the label declaration org/space specific? As far as I can see from the API spec, querying for labels is independent of any other organizational entity (like org and space). I get the impression that we would want to have a "second discovery mechanism" - very much like we did in #61.

Additionally, what is the use case that you have in your mind saying that you want to declare the exact match of the tag yourself in Promregator's config? Why wouldn't it be sufficient -- let's say as an example -- that we hard-code a tag called ```metrics-monitoring: promregator```? If there might be more than just one tool we want to support, we could also say that the value of the tag could be a comma-separated list and as long as "promregator" is mentioned there, we take it...
Or do you have some special reason for doing it like you proposed it initially?

**Comment by eaglerainbow** [comment]
I am just reading the documentation at https://docs.cloudfoundry.org/adminguide/metadata.html again. There I get another idea on how we could configure this:

* Let's say that we would declare something like a "label group" in Prometheus' config. 
* Each label group gets an own name (in the initial suggestion this would be `spring-boot`)
* For each named label group, we define all the attributes necessary (e.g. path, protocol, ...)
* The applications on CF would get tagged with `metrics-group: <label-group-id>`.
* For such each application on CF, Promregator would lookup the `label-group-id` and check if there was a configuration provided. If yes, the "configuration pattern" there is applied.

Additionally, as labels not just can be added to apps but also to orgs and spaces, we also could (in a second iteration) apply it on org/space level, i.e. if a space was tagged accordingly, then all apps of that space would be considered as target (applying the "config pattern").

Any ideas about that?

**Comment by martin-schaefer** [comment]
> Apparently, there is no CF CLI support for it, but you must `cf curl` it manually.
Yes that is true. I took a look at this today and labels/annotations are supported on the CF REST API and also in the Pivotal UI. (See attached screenshot). So its no problem to use it, even when missing in the CF cli.

![Pivotal Metadata](https://user-images.githubusercontent.com/11317256/58506280-3326d480-818f-11e9-99b5-9726b23737d7.png)

**Comment by eaglerainbow** [comment]
There is also one additional prerequisite for the entire endeavor: "cf-client-java" (the underlying library, which Promregator uses to talk to the CF platform) must be able to support it. If it does not, we first would have to request it from upstream. 

The only thing I found so far in this regard is https://github.com/cloudfoundry/cf-java-client/blob/92c20021df2b997bb1d9bbd6af28df3761ae4165/cloudfoundry-operations/src/main/java/org/cloudfoundry/operations/serviceadmin/DefaultServiceAdmin.java#L403 , which isn't too much telling right now... So, we would have to test that first.

**Comment by eaglerainbow** [comment]
@martin-schaefer One more aspect that just comes to my mind on this: Are users having assigned only the role "auditor" capable of querying/searching for labels? If they are, fine. If they don't, then we run in a similar ugly discussion as we did with UPSes (in #61).

**Comment by martin-schaefer** [comment]
> Why do you want to have the label declaration org/space specific?
That was not my idea, but my example was probably misleading: As all other filter levels (orgunit, space, app name) labels could be used standalone or together with one or more of the others. Its just one more, optional criteria for the folter, So the following target configs would all be valid:

```yaml
     targets:
       - labels: 
           metrics-provider-type: spring-boot
         path: actuator/prometheus
       - labels: 
           metrics-provider-type: standard
       - orgName: my-org
         labels: 
           metrics-provider-type: spring-boot
         path: actuator/prometheus
       - orgName: my-org
         spaceName: my-space
         labels: 
           metrics-provider-type: homegrown
         path: homegrown/metrics
```

**Comment by eaglerainbow** [comment]
> labels/annotations are supported on the CF REST API and also in the Pivotal UI.

Note that while this may be true for Pivotal UI, it may not hold true for other cockpits of other CF providers. For example, I have access to IBM Cloud (f.k.a. Bluemix) and SAP Cloud Platform. In both cockpits I could not find a functionality for maintaining labels in neither of them.

**Comment by eaglerainbow** [comment]
>  So the following target configs would all be valid:

Ah, you more consider this as "another filter criteria to targets declaration".

Mmh... For example, the case

``` 
 targets:
       - labels: 
           metrics-provider-type: spring-boot
         path: actuator/prometheus
``` 
makes it look very ugly, as this would mean that we have `path` declarations on the top level - which feel to be standalone. Not just the fact that this would hell to implement such a configuration, it is also slang-ish and may lead to contradictions. 
For example, let's take the following variant of your initial proposal:

``` 
     targets:
       - labels: 
           metrics-provider-type: spring-boot
         path: actuator/prometheus
       - orgName: my-org
         labels: 
           metrics-provider-type: spring-boot
         path: somethingelse/prometheus
```

Which `path` will be used for apps in org `my-org`? Will we scrape them only once - or with both?
IMHO are such kind of configurations hell to both admins and developers... That's why I would like to avoid them (if possible).

**Comment by eaglerainbow** [comment]
In the configuration example
```yaml

       - orgName: my-org
         spaceName: my-space
         labels: 
           metrics-provider-type: homegrown
         path: homegrown/metrics
```

to which attribute (org or space, or app) would `labels` refer here?

**Comment by eaglerainbow** [comment]
> Its just one more, 

What do you exactly mean by that? Do you mean `label` as only in the sense of "label-on-application" in your examples?

**Comment by martin-schaefer** [comment]
>to which attribute (org or space, or app) would labels refer here?

To the app attribute, I think this will be sufficient,

>What do you exactly mean by that? Do you mean label as only in the sense of "label-on-application" in your examples?

Exactly ;-)

> I have access to IBM Cloud (f.k.a. Bluemix) and SAP Cloud Platform. In both cockpits I could not find a functionality for maintaining labels in neither of them.

But AFAIK it is part of the standard CF REST API, or not?
Another approach could be to use a service registry and do the lookup there. In CF it seems that Eureka is widely used (and not supported by Prometheus).

**Comment by martin-schaefer** [comment]
Probably the attribute should better be named **appLabels**

**Comment by eaglerainbow** [comment]
> Exactly ;-)

Okay, that would simply make it as an optional, additional filter condition like we have it in https://github.com/promregator/promregator/blob/8ab4d27a0a6415fd1ae43f1dd419d080d5a2c056/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java#L134

**Comment by eaglerainbow** [comment]
> Probably the attribute should better be named appLabels

Yeah, I would prefer this naming, too, as it would be much more precise and much less misleading. I most likely wouldn't have fallen for the different interpretation of it in first place.

**Comment by martin-schaefer** [comment]
Yes, sorry. I did not make the Intention clear. I had in mind to make it as simple as possible and to fit in the existing target attributes. No existing target attributes must be changed.

**Comment by eaglerainbow** [comment]
> > I have access to IBM Cloud (f.k.a. Bluemix) and SAP Cloud Platform. In both cockpits I could not find a functionality for maintaining labels in neither of them.

> But AFAIK it is part of the standard CF REST API, or not?

Yes, it is: They all have the (standard CAPI) CF REST API as well, but can only reach the stuff either by sending HTTP requests manually or by using `cf curl` (that's what I had meant initially). 

> Another approach could be to use a service registry and do the lookup there. In CF it seems that Eureka is widely used (and not supported by Prometheus).

That then ends up in having metadata in two places which are then not properly coupled in lifecycle. From my experience, this is just asking for trouble.
And if we want to have proper coupling to the lifecycle, we then end up almost immediately in the UPS discussion again. 😞

**Comment by eaglerainbow** [comment]
> Yes, sorry. I did not make the Intention clear. 

No need to say sorry! We are discussing here on what is the right requirement / concept for that enhancement. And yes, that might imply that two brains have to different (incompatible) paths that they try to walk down. It's much better to find such a misinterpretation right now than compared to if we have implemented it and released it to the world... that would become much more ugly.

**Comment by martin-schaefer** [comment]
Eureka support would need a additional target type (like prometheus has different scrape definitions) and is certainly not something done in 5 minutes.

**Comment by eaglerainbow** [comment]
> Eureka support [...] is certainly not something done in 5 minutes.

While I agree to that statement, it still does not matter, if it would take 5 minutes of 5 days to make it happen. If it was the Right Thing™️ and we all would be convinced that it would solve all the problems, then I would be with it. 
In short, it's allowed here to think "a little big" 😀

**Comment by eaglerainbow** [comment]
From all what I read so far here, I would summarize the state of requirement right now like this:

* There shall be another map-typed attribute in the list of `targets` called `appLabels`.
* A key-name value pair there provides another condition to the discovered apps specified before. That means: all apps discovered by other configuration options must *additionally* also have the appLabel provided. If the appLabel was missing (e.g. it isn't set), then the app is not discovered/ignored by Promregator.
* This implies that (for example) `orgName` and `spaceName` could be omitted; still the `appLabels` would only work on `application` level.
* There is no such thing as "labels on org" or "labels on space" level (yet). 

What comes up to my mind, when writing this right now:

* What happens if there are provided more than just one `appLabel` mapping? Are they ORed or ANDed? Or shall we only allow one?

I think it is worthwhile to have a brief prototype on this assumption to see where we end up with all this.
**If nobody else would like to volunteer on that**, I would give it a try within the next 3-5 days somewhere (assuming that I find time for it).

**Comment by martin-schaefer** [comment]
Thats it. Precisely in a nutshell.

**Comment by eaglerainbow** [comment]
We have our first answer:

> Is cf-java-client able to query for labels on app level?

It does not support it (yet). For the case of `ListServicesRequest`s, cf-java-client is able to do so due to https://github.com/cloudfoundry/cf-java-client/blob/92c20021df2b997bb1d9bbd6af28df3761ae4165/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v2/services/_ListServicesRequest.java#L42  which mainly is label support: 
```java
    /**
     * The labels
     */
    @FilterParameter("label")
    @Nullable
abstract List<String> getLabels();
```

For both v2 applications and v3 applications, label support is not available: The files at https://github.com/cloudfoundry/cf-java-client/blob/92c20021df2b997bb1d9bbd6af28df3761ae4165/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v2/applications/_ListApplicationsRequest.java and https://github.com/cloudfoundry/cf-java-client/blob/92c20021df2b997bb1d9bbd6af28df3761ae4165/cloudfoundry-client/src/main/java/org/cloudfoundry/client/v3/applications/_ListApplicationsRequest.java respectively do not have such a 

```java
 @FilterParameter("label")
```

**Comment by eaglerainbow** [comment]
BTW: https://docs.cloudfoundry.org/adminguide/metadata.html states nothing about which API version (v2/v3) must be used for specifying labels. Given the examples, I would assume that only v3 applications will be supported.
That brings us into another dilemma: Downwards-compatibility! We must make sure that Promregator does not break, if only an old API version (using v2 applications only) is available. 
👉 another requirement in our list

**Comment by eaglerainbow** [comment]
I have raised https://github.com/cloudfoundry/cf-java-client/issues/967 asking for label support for apps.

**Comment by eaglerainbow** [comment]
> BTW: https://docs.cloudfoundry.org/adminguide/metadata.html states nothing about which API version (v2/v3) must be used for specifying labels. Given the examples, I would assume that only v3 applications will be supported.

I just ran a brief test. On Bluemix the request

``` bash
$ cf curl v2/apps/GUIDGOESHERE -X PATCH \
-d '{ "metadata": { "labels": { "test": "value" } } }'
```

ended up in

``` 
{
   "description": "Unknown request",
   "error_code": "CF-NotFound",
   "code": 10000
}
```

Note that 
``` bash
$ cf api
api endpoint:   https://api.eu-gb.bluemix.net
api version:    2.128.0
```

In a nutshell: We must make use of v3 applications for that feature :dizzy_face: So, we end up in the downards-compatibility discussion.

**Comment by eaglerainbow** [comment]
Here is another answer:

> Does querying labels work, if the user only has the role "Auditor" (maintaining labels is a different story and is out of scope)?

I just created a small test case on an SCP platform (running CF API version 2.135.0). There, retrieving the label of an application and querying all applications for a given label worked -- even if the user only had the "Auditor" role for the corresponding space.

👉 That should be okay and not create any additional requirement for us.

**Comment by eaglerainbow** [comment]
Ad 

> How can we handle map-typed attributes in Spring configuration properties? (I don't know this by heart)

Along the lines of https://stackoverflow.com/a/31051377, the extended configuration

```yaml

promregator:
  testmap:
    a: b
```

could be read with an extension of class `PromregatorConfiguration` like this:

```java
	private Map<String, String> testmap;

	public Map<String, String> getTestmap() {
		return this.testmap;
	}
	
	public void setTestmap(Map<String, String> map) {
		this.testmap = map;
	}
``` 
Note that both the setter and the getter method is important here! If they are not provided, value injection does not work for spring.

**Comment by eaglerainbow** [comment]
## Intermediate Summary of Status 

1. For item 1 we detected that there is no support by cf-java-client. Support has been requested with https://github.com/cloudfoundry/cf-java-client/issues/967
2. Map-typed configuration properties are possible in spring-framework. Details can be found at https://github.com/promregator/promregator/issues/110#issuecomment-497098118
3. Given that item 1 is not resolved, tackling item 3 is not possible yet.
4. The "auditor" role is sufficient for retrieving and querying labels.

**Comment by martin-schaefer** [comment]
Thanks for the in-depth analysis. My point-of-view for the three questions:

- This class might be of help: https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/client/CloudFoundryClientCompatibilityChecker.java
- I would log a warning when v3 is not available and ignore the labels.
- I would use ANDed because other criteria is also ANDed and therefore it will be consistent. Even though I think the case for using several lables seems not very likely.

**Comment by eaglerainbow** [comment]
Thanks for your ideas + the pointer to `CloudFoundryClientCompatibilityChecker`.

>  Even though I think the case for using several lables seems not very likely.

Given that statement, would it be okay to not have a map for the labels, but to only permit one single combination of labelName+labelValue?

**Comment by braunsonm** [comment]
The annotations supported filtering will help with this. Though filtering by annotations is better since it does the filtering on CAPI instead of in promregator, annotations is more mainstream with what other Cloud Foundry projects are doing it seems.

Eitherway adding labels support will be easier now that parts of Promregator are using V3 APIs
